src/NuGet.Core/NuGet.Protocol.Core.v3/Utility/LocalFolderUtility.cs

```diff
@@ -0,0 +1,685 @@
+
+                // Check for any files that are in the form packageId.version.nupkg
+                foreach (var file in directoryList.Where(file => IsPossiblePackageMatch(file, identity)))
+                {
+                    var package = GetPackageFromNupkg(file);
+
+                    if (identity.Equals(package.Identity))
+                    {
+                        if (StringComparer.OrdinalIgnoreCase.Equals(
+                            identity.Version.ToString(),
+                            package.Identity.Version.ToString()))
+                        {
+                            // Take an exact match immediately
+                            return package;
+                        }
+                        else if (fallbackMatch == null)
+                        {
+                            // This matches the identity, but there may be an exact match still
+                            fallbackMatch = package;
+                        }
+                    }
+                }
+
+                if (fallbackMatch != null)
+                {
+                    // Use the fallback match if an exact match was not found
+                    return fallbackMatch;
+                }
+            }
+
+            // Not found
+            return null;
+        }
+
+        /// <summary>
+        /// True if the file name matches the identity. This is could be incorrect if
+        /// the package name ends with numbers. The result should be checked against the nuspec.
+        /// </summary>
+        public static bool IsPossiblePackageMatch(FileInfo file, PackageIdentity identity)
+        {
+            if (identity == null)
+            {
+                throw new ArgumentNullException(nameof(identity));
+            }
+
+            if (file == null)
+            {
+                throw new ArgumentNullException(nameof(file));
+            }
+
+            return identity.Equals(GetIdentityFromFile(file, identity.Id));
+        }
+
+        /// <summary>
+        /// True if the file name matches the id and is followed by a version. This is could be incorrect if
+        /// the package name ends with numbers. The result should be checked against the nuspec.
+        /// </summary>
+        public static bool IsPossiblePackageMatch(FileInfo file, string id)
+        {
+            if (id == null)
+            {
+                throw new ArgumentNullException(nameof(id));
+            }
+
+            if (file == null)
+            {
+                throw new ArgumentNullException(nameof(file));
+            }
+
+            return GetIdentityFromFile(file, id) != null;
+        }
+
+        /// <summary>
+        /// An imperfect attempt at finding the identity of a package from the file name.
+        /// This can fail if the package name ends with something such as .1
+        /// </summary>
+        public static PackageIdentity GetIdentityFromFile(FileInfo file, string id)
```

Comment:
This method is strange to me -- shouldn't it just return a NuGetVersion? Also, it should be documented (or even named) that this only works on .nupkgs paths. Also, since this only uses file.Name, could that be the parameter?